src/ring_theory/polynomial/chebyshev/basic.lean

```diff
@@ -0,0 +1,204 @@
+  { let K := fraction_ring (polynomial (zmod p)),
+    let f : zmod p →+* K := (fraction_ring.of _).to_map.comp C,
+    haveI : char_p K p := by { rw ← f.char_p_iff_char_p, apply_instance },
+    haveI : infinite K :=
+    by { apply infinite.of_injective _ (fraction_ring.of _).injective, apply_instance },
+    refine ⟨K, _, _, _⟩; apply_instance },
+  resetI,
+  apply map_injective (zmod.cast_hom (dvd_refl p) K) (ring_hom.injective _),
+  rw [map_lambdashev, map_pow, map_X],
+  apply eq_of_infinite_eval_eq,
+  -- The two polynomials agree on all `x` of the form `x = y + y⁻¹`.
+  apply @set.infinite_mono _ {x : K | ∃ y, x = y + y⁻¹ ∧ y ≠ 0},
+  { rintro _ ⟨x, rfl, hx⟩,
+    simp only [eval_X, eval_pow, set.mem_set_of_eq, @add_pow_char K _ p,
+      lambdashev_eval_add_inv _ _ (mul_inv_cancel hx)] },
+  -- Now we need to show that the set of such `x` is infinite.
+  -- If the set is finite, then we will show that `K` is also finite.
+  { intro h,
+    rw ← set.infinite_univ_iff at H,
+    apply H,
+    -- To each `x` of the form `x = y + y⁻¹`
+    -- we `bind` the set of `y` that solve the equation `x = y + y⁻¹`.
+    -- For every `x`, that set is finite (since it is governed by a quadratic equation).
+    -- For the moment, we claim that all these sets together cover `K`.
+    suffices : (set.univ : set K) =
+      {x : K | ∃ (y : K), x = y + y⁻¹ ∧ y ≠ 0} >>= (λ x, {y | x = y + y⁻¹ ∨ y = 0}),
+    { rw this, clear this,
+      apply set.finite_bind h,
+      rintro x hx,
+      -- The following quadratic polynomial has as solutions the `y` for which `x = y + y⁻¹`.
+      let φ : polynomial K := X ^ 2 - C x * X + 1,
+      have hφ : φ ≠ 0,
+      { intro H,
+        have : φ.eval 0 = 0, by rw [H, eval_zero],
+        simpa [eval_X, eval_one, eval_pow, eval_sub, sub_zero, eval_add,
+          eval_mul, mul_zero, pow_two, zero_add, one_ne_zero] },
+      classical,
+      convert (φ.roots ∪ {0}).to_finset.finite_to_set using 1,
+      ext1 y,
+      simp only [multiset.mem_to_finset, set.mem_set_of_eq, finset.mem_coe, multiset.mem_union,
+        mem_roots hφ, is_root, eval_add, eval_sub, eval_pow, eval_mul, eval_X, eval_C, eval_one,
+        multiset.mem_singleton, multiset.singleton_eq_singleton],
+      by_cases hy : y = 0,
+      { simp only [hy, eq_self_iff_true, or_true] },
+      apply or_congr _ iff.rfl,
+      rw [← mul_left_inj' hy, eq_comm, ← sub_eq_zero, add_mul, inv_mul_cancel hy],
+      apply eq_iff_eq_cancel_right.mpr,
+      ring },
+    -- Finally, we prove the claim that our finite union of finite sets covers all of `K`.
+    { apply (set.eq_univ_of_forall _).symm,
+      intro x,
+      simp only [exists_prop, set.mem_Union, set.bind_def, ne.def, set.mem_set_of_eq],
+      by_cases hx : x = 0,
+      { simp only [hx, and_true, eq_self_iff_true, inv_zero, or_true],
+        exact ⟨_, 1, rfl, one_ne_zero⟩ },
+      { simp only [hx, or_false, exists_eq_right],
+        exact ⟨_, rfl, hx⟩ } } }
+end
```

Comment:
It's so cool that you can actually carry out this argument without a too-unpleasant de Bruijn factor!

Reply:
Yes, I was very pleased by this myself.